hspec-core/test/Test/Hspec/Core/SpecSpec.hs

```diff
@@ -20,6 +20,30 @@ spec :: Spec
 spec = do
   let runSpecM = fmap snd . H.runSpecM
 
+  describe "askAncestors" $ do 
+    it "works" $ do 
+      res <- runSpec $ do 
+        ancs0 <- H.askAncestors
+        H.it "works at root" $ do
+          ancs0 `shouldBe` []
+        H.describe "child1" $ do
+          ancs1 <- H.askAncestors
+          H.it "works with one-level tree" $ do
+            ancs1 `shouldBe` ["child1"]
+          H.describe "grandchild1" $ do
+            ancs2 <- H.askAncestors
+            H.it "works with two-level tree" $ do
+              ancs2 `shouldBe` ["grandchild1", "child1"]
+          H.describe "grandchild2" $ do
+            ancs2 <- H.askAncestors
+            H.it "works with two-level tree" $ do
+              ancs2 `shouldBe` ["grandchild2", "child1"]
+        H.describe "child2" $ do
+          ancs1 <- H.askAncestors
+          H.it "works with one-level tree" $ do
+            ancs1 `shouldBe` ["child2"]
+      last res `shouldBe` "5 examples, 0 failures"
```

Comment:
Most of the work is done by `ReaderT` and `local`. I think if we got the basic wiring  right it’s very hard to have a bug here. What I think is, one simple test case will be enough. I think we can also test it directly, without using `runSpec`, something like:

```
void . runSpecM $ do
  describe …
    describe…
      askAncestors >>= runIO . (`shouldBe` […])
```

Reply:
I didn't know about `runIO`; looks like it is something to use when evaluating a test without `it`/`describe`? I've reduced the tests.

Reply:
Not exactly. `runIO` runs some IO during spec construction time. This is why we can use it to verify that `getSpecDescriptionPath` returns the right thing during spec construction.